..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/

---   
 <img align="left" width="75" height="75"  src="https://upload.wikimedia.org/wikipedia/en/c/c8/University_of_the_Punjab_logo.png"> 

<h1 align="center">Department of Data Science</h1>
<h1 align="center">Course: Tools and Techniques for Data Science</h1>

---
<h3><div align="right">Instructor: Muhammad Arif Butt, Ph.D.</div></h3>    

<h1 align="center">Lecture 3.12 (Pandas-04)</h1>

---   
 <img align="left" width="75" height="75"  src="https://upload.wikimedia.org/wikipedia/en/c/c8/University_of_the_Punjab_logo.png"> 

<h1 align="center">Department of Data Science</h1>
<h1 align="center">Course: Tools and Techniques for Data Science</h1>

---
<h3><div align="right">Instructor: Muhammad Arif Butt, Ph.D.</div></h3>    

## _IO with CSV EXCEL and JSON Files_

<img align="center" width="600" height="150"  src="..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/images/fileformats.png" >

## Learning agenda of this notebook
[Pandas](https://pandas.pydata.org/) provides helper functions to read data from various file formats like CSV, EXCEL, JSON, HTML, SQL table, and many more.
1. Reading data from a CSV/TSV File
2. Reading a CSV file from a Remote System
3. Writing Contents of Dataframe to a CSV File
4. Reading data from an EXCEL File
5. Writing Contents of Dataframe to an EXCEL File
6. Reading data from a JSON File
7. Writing Contents of Dataframe to a JSON File

In [6]:
# import sys
# !{sys.executable} -m pip install panda --quiet

In [7]:
import pandas as pd
pd.__version__, pd.__path__

('2.0.3', ['C:\\Users\\FashN\\anaconda3\\Lib\\site-packages\\pandas'])

## 1. Reading from CSV/TSV Files
>**CSV**: A text file in which the values are separated by a comma or a tab character is called a CSV or a TSV file. Each line of the file is a data record and each record consists of one or more fields, separated by a specific character called separator. A CSV/TSV file is typically used to store tabular data (numbers and text), in which each line will have the same number of fields.

### a. Reading a  Simple CSV File
The `pd.read_csv()` method is used to read a comma-separated file into a DataFrame.
```
pd.read_csv(fname, delimiter=None, header='infer', skiprows=None , nrows=None , usecols=None,  footer='',...)

```

In [8]:
! cat datasets/classmarks.csv

'cat' is not recognized as an internal or external command,
operable program or batch file.


In [11]:
#The `read_csv`, by default assumes that the file contains comma separated values, 
# and the first row of the file conatins names of columns, which will be taken as column labels

df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarks.csv')
df

rollno  gender    group   age  math  english  urdu
0    MS01  female  group B  28.0  72.0       72  74.0
1    MS02  female  group C  33.0  69.0       90  88.0
2    MS03  female  group B  21.0   NaN       95  93.0
3    MS04    male  group A  44.0  47.0       57  44.0
4    MS05    male  group C  54.0  76.0       78   NaN
5    MS06  female  group B   NaN  71.0       83  78.0
6    MS07  female  group B  47.0  88.0       95  92.0
7    MS08    male  group B  33.0  40.0       43  39.0
8    MS09    male  group D  27.0  64.0       64  67.0
9    MS10  female  group B  33.0  38.0       60  50.0
10   MS11    male  group C  46.0  58.0       54  52.0
11   MS12    male  group D  53.0  40.0       52  43.0
12   MS13  female  group B  25.0  65.0       81  73.0
13   MS14    male  group A  26.0  78.0       72  70.0
14   MS15  female  group A  25.0  50.0       53  58.0
15   MS16  female  group C  29.0  69.0       75  78.0
16   MS17    male  group C  22.0  88.0       89  86.0
17   MS18  female  group B  31.0  18.0       32  28.0
18   MS19    male  group C  33.0  46.0       42  46.0
19   MS20  female  group C  36.0  54.0       58   NaN
20   MS21    male  group D  34.0  66.0       69  63.0
21   MS22  female  group B  32.0  65.0       75  70.0
22   MS23    male  group D  31.0  44.0       54  53.0
23   MS24  female  group C  37.0   NaN       73  73.0
24   MS25    male  group D  33.0  74.0       71  80.0
25   MS26    male  group A  39.0  73.0       74  72.0
26   MS27    male  group B   NaN  69.0       54  55.0
27   MS28  female  group C  47.0  67.0       69  75.0
28   MS29    male  group C  44.0  70.0       70  65.0
29   MS30  female  group D  42.0  62.0       70  75.0
30   MS31  female  group D  41.0  69.0       74  74.0
31   MS32  female  group B   NaN  63.0       65  61.0
32   MS33  female  group E  29.0  56.0       72  65.0
33   MS34    male  group D  34.0  40.0       42  38.0
34   MS35    male  group E  45.0  97.0       87  82.0
35   MS36    male  group E  31.0  81.0       81  79.0
36   MS37  female  group D  28.0  74.0       81   NaN
37   MS38  female  group D  26.0  50.0       64  59.0
38   MS39  female  group D  36.0  75.0       90  88.0
39   MS40    male  group B  33.0  57.0       56  57.0
40   MS41    male  group C  35.0  55.0       61  54.0
41   MS42  female  group C  22.0  58.0       73  68.0
42   MS43  female  group B  38.0  53.0       58  65.0
43   MS44    male  group B  31.0  59.0       65  66.0
44   MS45  female  group E  28.0   NaN       56  54.0
45   MS46    male  group B  26.0  65.0       54  57.0
46   MS47  female  group A  36.0  55.0       65  62.0
47   MS48  female  group C  30.0  66.0       71  76.0
48   MS49  female  group D  40.0  57.0       74  76.0
49   MS50    male  group C  37.0  66.0       78  81.0

**The `df.head(N)` method is used to select/display first `N` rows, based on `position`, i.e., the integer value corresponding to the position of the row (from 0 to n-1). The default value of `N` is 5.**

In [12]:
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

In [13]:
df.tail()

rollno  gender    group   age  math  english  urdu
45   MS46    male  group B  26.0  65.0       54  57.0
46   MS47  female  group A  36.0  55.0       65  62.0
47   MS48  female  group C  30.0  66.0       71  76.0
48   MS49  female  group D  40.0  57.0       74  76.0
49   MS50    male  group C  37.0  66.0       78  81.0

In [14]:
df.head(3)

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0

In [16]:
df.head(-48)

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0

In [17]:
df.tail(-48)

rollno  gender    group   age  math  english  urdu
48   MS49  female  group D  40.0  57.0       74  76.0
49   MS50    male  group C  37.0  66.0       78  81.0

### b.Reading a CSV File having a Delimter, other than Comma
- By default, the `read_csv()` expect comma as seperator. But if the CSV file has some other seperator or delimiter like (semi-collon or tab), it will raise an error.
- To handler the issue we need to pass specific value to the `delimiter` argument of `read_csv()` method.

In [19]:
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithtab.csv')
df.head()

rollno\tgender\tgroup\tage\tmath\tenglish\turdu
0     MS01\tfemale\tgroup B\t28.0\t72.0\t72\t74.0
1     MS02\tfemale\tgroup C\t33.0\t69.0\t90\t88.0
2         MS03\tfemale\tgroup B\t21.0\t\t95\t93.0
3       MS04\tmale\tgroup A\t44.0\t47.0\t57\t44.0
4           MS05\tmale\tgroup C\t54.0\t76.0\t78\t

In [20]:
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithtab.csv', delimiter='\t', header=None)
df.head()

0       1        2     3     4        5     6
0  rollno  gender    group   age  math  english  urdu
1    MS01  female  group B  28.0  72.0       72  74.0
2    MS02  female  group C  33.0  69.0       90  88.0
3    MS03  female  group B  21.0   NaN       95  93.0
4    MS04    male  group A  44.0  47.0       57  44.0

### c. Reading a CSV File not having Column Labels
- By default the `read_csv()` method assume the first row of the file will contain column labels
- If this is not the case, i.e., the file do not contain column labels rather data, it will be dealt as column label
- Understand this in following example

In [23]:
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithoutcollabels.csv')
df.head()

MS01  female  group B    28    72  72.1    74
0  MS02  female  group C  33.0  69.0    90  88.0
1  MS03  female  group B  21.0   NaN    95  93.0
2  MS04    male  group A  44.0  47.0    57  44.0
3  MS05    male  group C  54.0  76.0    78   NaN
4  MS06  female  group B   NaN  71.0    83  78.0

**To read such files, you have to pass the parameter `header=None` to the `read_csv()` method as shown below**

In [24]:
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithoutcollabels.csv', header=None)
df.head()

0       1        2     3     4   5     6
0  MS01  female  group B  28.0  72.0  72  74.0
1  MS02  female  group C  33.0  69.0  90  88.0
2  MS03  female  group B  21.0   NaN  95  93.0
3  MS04    male  group A  44.0  47.0  57  44.0
4  MS05    male  group C  54.0  76.0  78   NaN

**Now if you want to assign new column labels to make them more understandable, you can assign the list of column labels to the `columns` attribute of the dataframe object**

In [26]:
col_names = ['rollno', 'gender', 'group','age', 'Math', 'English', 'Physics']
df.columns = col_names
df.head()

rollno  gender    group   age  Math  English  Physics
0   MS01  female  group B  28.0  72.0       72     74.0
1   MS02  female  group C  33.0  69.0       90     88.0
2   MS03  female  group B  21.0   NaN       95     93.0
3   MS04    male  group A  44.0  47.0       57     44.0
4   MS05    male  group C  54.0  76.0       78      NaN

### d. Reading a CSV File having Comments in the beginning
- You may get an error while reading a CSV file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the csv file in the text editor and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginninghaving comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [27]:
# Try reading a csv file having 3 comments lines in the beginning.
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithtopcomments.csv')
df.head()

These are comments Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0  These are second line of comments        NaN        NaN        NaN   
1   These are third line of comments        NaN        NaN        NaN   
2                             rollno     gender      group        age   
3                               MS01     female    group B         28   
4                               MS02     female    group C         33   

  Unnamed: 4 Unnamed: 5 Unnamed: 6  
0        NaN        NaN        NaN  
1        NaN        NaN        NaN  
2       math    english       urdu  
3         72         72         74  
4         69         90         88

In [28]:
# Try reading a csv file having 3 comments lines in the beginning.
df = pd.read_csv('..\..\..\data_science\exercise\Section-3-Python-for-Data-Scientists/datasets/classmarkswithtopcomments.csv', skiprows=3)
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

### e. Reading a portion of CSV File in a Dataframe
- Suppose the dataset inside the csv file is too big and you don't want to spend that much time for reading that data
- Or might be your system crashes, when you try to load that much data
- Solution is read
    - Specific number of rows by passing `nrows` parameter to `read_csv()` method
    - Specific number of columns by passing `usecols` parameter to `read_csv()` method

In [33]:
df = pd.read_csv('datasets/classmarks.csv', nrows=3)
print(f'shape = {df.shape}')
df.head()

shape before = (3, 7)


rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0

In [34]:
# Read specific columns from the csv file by passing a list of column names to `usecols` argument
df = pd.read_csv('datasets/classmarks.csv', usecols = ['rollno', 'age', 'math', 'english'])

df.shape

(50, 4)

In [35]:
df.head()

rollno   age  math  english
0   MS01  28.0  72.0       72
1   MS02  33.0  69.0       90
2   MS03  21.0   NaN       95
3   MS04  44.0  47.0       57
4   MS05  54.0  76.0       78

In [36]:
# Ofcourse you can use both the parameters at the same time
pd = pd.read_csv('datasets/classmarks.csv', nrows = 7, usecols=['rollno', 'group', 'english'])
df.shape

(50, 4)

In [38]:
df.head()

rollno   age  math  english
0   MS01  28.0  72.0       72
1   MS02  33.0  69.0       90
2   MS03  21.0   NaN       95
3   MS04  44.0  47.0       57
4   MS05  54.0  76.0       78

## 2. Reading a CSV File from a Remote System

In [40]:
# To avoid URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED]..... 
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

### a. Reading a CSV file from GitHub Gist

In [42]:
import pandas as pd
myurl = 'https://bit.ly/31eYGTx'
df = pd.read_csv(myurl)
df

Names  Age       Addr
0     Arif   50     Lahore
1     Rauf   52  Islamabad
2     Maaz   27   Peshawer
3   Hadeed   22  Islamabad
4  Mujahid   18    Karachi

In [43]:
sheetID = '1H9ZTGVRXN3zuyP3cbJQnbX7JlqhqFNMOE_WwDP-PRTE'
sheetName = 'sheet1'
URL = 'https://docs.google.com/spreadsheets/d/{0}/gviz/tq?tqx=out:csv&sheet={1}'.format(sheetID, sheetName)

df = pd.read_csv(URL)
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

In [44]:
URL

'https://docs.google.com/spreadsheets/d/1H9ZTGVRXN3zuyP3cbJQnbX7JlqhqFNMOE_WwDP-PRTE/gviz/tq?tqx=out:csv&sheet=sheet1'

## 3. Writing Contents of Dataframe to a CSV File
- The `pd.to_csv()` method is used to write the contents of a dataframe (with indices) to a CSV file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

In [46]:
df_class = pd.read_csv('datasets/classmarks.csv')
df_class.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

>- Let us create a new dataframe from above dataframe containing records of only group B

In [47]:
mask = (df_class['group'] == 'group B')
mask.head()

0     True
1    False
2     True
3    False
4    False
Name: group, dtype: bool

In [52]:
gender = (df_class['gender'] == 'female')
gender.head()

0     True
1     True
2     True
3    False
4    False
Name: gender, dtype: bool

In [50]:
df_class_groupB = df_class.loc[mask]
df_class_groupB.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
2   MS03  female  group B  21.0   NaN       95  93.0
5   MS06  female  group B   NaN  71.0       83  78.0
6   MS07  female  group B  47.0  88.0       95  92.0
7   MS08    male  group B  33.0  40.0       43  39.0

In [54]:
df_class_gender = df_class.loc[gender]
df_class_gender.tail()

rollno  gender    group   age  math  english  urdu
42   MS43  female  group B  38.0  53.0       58  65.0
44   MS45  female  group E  28.0   NaN       56  54.0
46   MS47  female  group A  36.0  55.0       65  62.0
47   MS48  female  group C  30.0  66.0       71  76.0
48   MS49  female  group D  40.0  57.0       74  76.0

In [57]:
df_class_groupB.to_csv('masks/classmarksgroupB.csv')

In [56]:
df_class_gender.to_csv('masks/classmarksGender.csv')

In [62]:
df = pd.read_csv('masks/classmarksGender.csv')
df.head()

Unnamed: 0 rollno  gender    group   age  math  english  urdu
0           0   MS01  female  group B  28.0  72.0       72  74.0
1           1   MS02  female  group C  33.0  69.0       90  88.0
2           2   MS03  female  group B  21.0   NaN       95  93.0
3           5   MS06  female  group B   NaN  71.0       83  78.0
4           6   MS07  female  group B  47.0  88.0       95  92.0

>To avoid writing the row indices column inside the file pass `index=False` argument to `to_csv()` method

In [65]:
df_class_groupB.to_csv('datasets/classmarksgroupB.csv', index=False)

In [66]:
df = pd.read_csv('datasets/classmarksgroupB.csv')
df

rollno  gender    group   age  math  english  urdu
0    MS01  female  group B  28.0  72.0       72  74.0
1    MS03  female  group B  21.0   NaN       95  93.0
2    MS06  female  group B   NaN  71.0       83  78.0
3    MS07  female  group B  47.0  88.0       95  92.0
4    MS08    male  group B  33.0  40.0       43  39.0
5    MS10  female  group B  33.0  38.0       60  50.0
6    MS13  female  group B  25.0  65.0       81  73.0
7    MS18  female  group B  31.0  18.0       32  28.0
8    MS22  female  group B  32.0  65.0       75  70.0
9    MS27    male  group B   NaN  69.0       54  55.0
10   MS32  female  group B   NaN  63.0       65  61.0
11   MS40    male  group B  33.0  57.0       56  57.0
12   MS43  female  group B  38.0  53.0       58  65.0
13   MS44    male  group B  31.0  59.0       65  66.0
14   MS46    male  group B  26.0  65.0       54  57.0

## 4. I/O with EXCEL Files
>**XLSX**: XLSX is a Microsoft Excel Open XML file format. It also comes under the Spreadsheet file format. It is an XML-based file format created by Microsoft Excel. In XLSX data is organized under the cells and columns in a sheet. Each XLSX file may contain one or more sheets. So a workbook can contain multiple sheets

### a. Reading a Simple Excel File


In [67]:
df = pd.read_excel(io='datasets/classmarks.xlsx')
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

In [68]:
df.shape

(50, 7)

### b. Reading an Excel File having Comments in the beginning
- You may get an error while reading an Excel file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the Excel file in MS EXCEL and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginning having comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [69]:
# The following file has three lines of comments in the beginning of the file.
df = pd.read_excel(io="datasets/classmarkswithcomments.xlsx", skiprows=3)
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

### c. Reading Excel Workbook with Multiple Sheets
- By default `pd.read_excel()` function read only the first sheet.
- What if we want to read an Excel file having multiple sheets.
- The `big_mart_sales_with_multiple_sheets.xlsx` is a workbook that contains three sheets for different years data. The sheet names are 1985, 1987, and 1997

In [70]:
df = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx')
# if you check/view the data you can see, it only contains the data of first excel sheet (for the year 1985)
df.shape

(1463, 12)

In [72]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDP10          NaN          Low Fat         0.127470   
1           DRI11          NaN          Low Fat         0.034238   
2           FDW12          NaN          Regular         0.035400   
3           FDC37          NaN          Low Fat         0.057557   
4           FDC14          NaN          Regular         0.072222   

      Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
0   Snack Foods  107.7622            OUT027                       1985   
1   Hard Drinks  113.2834            OUT027                       1985   
2  Baking Goods  144.5444            OUT027                       1985   
3  Baking Goods  107.6938            OUT019                       1985   
4        Canned   43.6454            OUT019                       1985   

  Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0      Medium               Tier 3  Supermarket Type3          4022.7636  
1      Medium               Tier 3  Supermarket Type3          2303.6680  
2      Medium               Tier 3  Supermarket Type3          4064.0432  
3       Small               Tier 1      Grocery Store           214.3876  
4       Small               Tier 1      Grocery Store           125.8362

In [79]:
df_1985 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx', sheet_name = '1985')
df_1997 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx', sheet_name = '1997')
df_1987 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx', sheet_name = '1987')


In [80]:
print(f'1985 = {df_1985.shape}\n1997 = {df_1997.shape}\n1987 = {df_1987.shape}')

1985 = (1463, 12)
1997 = (930, 12)
1987 = (932, 12)


## 5. Writing Contents of Dataframe to an EXCEL File
- The `pd.to_excel()` method is used to write the contents of a dataframe (with indices) to an Excel file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

>- Let us create a new single dataframe after concatenating all the above three dataframes using `pd.concat()` method

In [82]:
df_concatenated = pd.concat(objs=[df_1985, df_1987, df_1997])
df_concatenated.shape

(3325, 12)

**Note the total number of rows in this dataframe equals to `1463+932+930 = 3325`**

In [83]:
df_concatenated.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDP10          NaN          Low Fat         0.127470   
1           DRI11          NaN          Low Fat         0.034238   
2           FDW12          NaN          Regular         0.035400   
3           FDC37          NaN          Low Fat         0.057557   
4           FDC14          NaN          Regular         0.072222   

      Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
0   Snack Foods  107.7622            OUT027                       1985   
1   Hard Drinks  113.2834            OUT027                       1985   
2  Baking Goods  144.5444            OUT027                       1985   
3  Baking Goods  107.6938            OUT019                       1985   
4        Canned   43.6454            OUT019                       1985   

  Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0      Medium               Tier 3  Supermarket Type3          4022.7636  
1      Medium               Tier 3  Supermarket Type3          2303.6680  
2      Medium               Tier 3  Supermarket Type3          4064.0432  
3       Small               Tier 1      Grocery Store           214.3876  
4       Small               Tier 1      Grocery Store           125.8362

In [84]:
# you can store the concatenated data inside your dataframe into a single Excel file
# You can mention the argument `index= false` for not storing row indices (0, 1,2,3,... in the Excel file.

df_concatenated.to_excel(excel_writer='temp.xlsx', index=False)
# Let us verify

data = pd.read_excel(io='temp.xlsx')

data

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDP10          NaN          Low Fat         0.127470   
1              DRI11          NaN          Low Fat         0.034238   
2              FDW12          NaN          Regular         0.035400   
3              FDC37          NaN          Low Fat         0.057557   
4              FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
3320           FDC28        7.905          Low Fat         0.054987   
3321           FDB17       13.150          Low Fat         0.036672   
3322           FDT34        9.300          Low Fat         0.174350   
3323           FDF53       20.750              reg         0.083607   
3324           DRG01       14.800          Low Fat         0.044878   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
0      Snack Foods  107.7622            OUT027                       1985   
1      Hard Drinks  113.2834            OUT027                       1985   
2     Baking Goods  144.5444            OUT027                       1985   
3     Baking Goods  107.6938            OUT019                       1985   
4           Canned   43.6454            OUT019                       1985   
...            ...       ...               ...                        ...   
3320  Frozen Foods  108.6254            OUT046                       1997   
3321  Frozen Foods  182.5976            OUT046                       1997   
3322   Snack Foods  104.4964            OUT046                       1997   
3323  Frozen Foods  178.8318            OUT046                       1997   
3324   Soft Drinks   75.4670            OUT046                       1997   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0         Medium               Tier 3  Supermarket Type3          4022.7636  
1         Medium               Tier 3  Supermarket Type3          2303.6680  
2         Medium               Tier 3  Supermarket Type3          4064.0432  
3          Small               Tier 1      Grocery Store           214.3876  
4          Small               Tier 1      Grocery Store           125.8362  
...          ...                  ...                ...                ...  
3320       Small               Tier 1  Supermarket Type1          1844.9318  
3321       Small               Tier 1  Supermarket Type1          4165.2448  
3322       Small               Tier 1  Supermarket Type1          2419.5172  
3323       Small               Tier 1  Supermarket Type1          3608.6360  
3324       Small               Tier 1  Supermarket Type1           765.6700  

[3325 rows x 12 columns]

In [85]:
data.shape

(3325, 12)

## 6. I/O with JSON Files

>**JSON**: JavaScript Object Notation is a text-based open standard file format that uses human-readable text consisting of attribute–value pairs and arrays. It is a data interchange format that is used to store and transfer the data via Internet, primarily between a web client and a server.

### a. Reading a Simple JSON File

In [86]:
# read the json file using read_json method of pandas library
df = pd.read_json('datasets/simple.json')
df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B

### b. Reading JSON File having each record in a separate line
- Some of the json files are written as records i.e each json line is a separate json object. For example:
```
{ 'name' : 'Ahsan', 'roll_no' : '100' } # line 1
{ 'name' : 'Ayesha' , 'roll_no' : '101' } # line 2
```

In [87]:
# To read such file you need to pass `lines=True` to the `read_json()` method of dataframe
df = pd.read_json('datasets/simple_records.json', lines=True)

df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B

## 7. Writing Contents of Dataframe to a JSON File

In [88]:
df.to_json('masks/temp.json')

pd.read_json('masks/temp.json')
df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B